File: src/python/xerxes/runtime/insights.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Analytics over recent :class:`CostEvent` records for the ``/insights`` command.

Aggregates a stream of :class:`CostEvent` entries (typically from a
:class:`CostTracker`) into daily totals, per-model and per-label breakdowns,
projected monthly burn, and cache savings. Tests inject synthetic events so
no live :class:`CostTracker` is needed.
"""

from __future__ import annotations

from collections.abc import Iterable
from dataclasses import dataclass, field
from datetime import UTC, datetime, timedelta

from .cost_tracker import CostEvent


def _parse_dt(s: str) -> datetime:
    """Parse ISO-8601 ``s`` to ``datetime``, falling back to now-UTC on failure."""
    try:
        return datetime.fromisoformat(s)
    except ValueError:
        return datetime.now(UTC)


@dataclass
class InsightsReport:
    """Aggregated stats over a list of :class:`CostEvent` records.

    Attributes:
        total_events: Number of events that contributed to this report.
        total_cost_usd: Sum of ``cost_usd`` across events.
        total_input_tokens: Sum of non-cached input tokens.
        total_output_tokens: Sum of output tokens.
        total_cache_read_tokens: Sum of cache-hit input tokens.
        total_cache_creation_tokens: Sum of cache-write tokens.
        by_model: Per-model rollup with events/cost/in/out token totals.
        by_day: Per-day rollup with event count and cost.
        by_label: Event-count histogram by ``label``.
        cache_hit_rate: Fraction of input tokens served from the cache.
        projected_monthly_cost: ``avg_daily_cost * 30`` extrapolation.
    """

    total_events: int = 0
    total_cost_usd: float = 0.0
    total_input_tokens: int = 0
    total_output_tokens: int = 0
    total_cache_read_tokens: int = 0
    total_cache_creation_tokens: int = 0
    by_model: dict[str, dict[str, float]] = field(default_factory=dict)
    by_day: dict[str, dict[str, float]] = field(default_factory=dict)
    by_label: dict[str, int] = field(default_factory=dict)
    cache_hit_rate: float = 0.0
    projected_monthly_cost: float = 0.0


def build_report(events: Iterable[CostEvent], *, days: int | None = None, now: datetime | None = None) -> InsightsReport:
    """Build an :class:`InsightsReport` from ``events``, optionally clipped to the last ``days``.

    Args:
        events: Iterable of :class:`CostEvent` records.
        days: When set, only events from the last ``days`` are included.
        now: Override clock for ``days`` filtering (defaults to UTC now).
    """
    horizon = (now or datetime.now(UTC)) - timedelta(days=days) if days else None
    rpt = InsightsReport()
    for ev in events:
        ts = _parse_dt(ev.timestamp)
        if ts.tzinfo is None:
            ts = ts.replace(tzinfo=UTC)
        if horizon is not None and ts < horizon:
            continue
        rpt.total_events += 1
        rpt.total_cost_usd += ev.cost_usd
        rpt.total_input_tokens += ev.in_tokens
        rpt.total_output_tokens += ev.out_tokens
        rpt.total_cache_read_tokens += ev.cache_read_tokens
        rpt.total_cache_creation_tokens += ev.cache_creation_tokens
        model_bucket = rpt.by_model.setdefault(ev.model, {"events": 0, "cost_usd": 0.0, "in_tokens": 0, "out_tokens": 0})
        model_bucket["events"] += 1
        model_bucket["cost_usd"] += ev.cost_usd
        model_bucket["in_tokens"] += ev.in_tokens
        model_bucket["out_tokens"] += ev.out_tokens
        day_key = ts.date().isoformat()
        day_bucket = rpt.by_day.setdefault(day_key, {"events": 0, "cost_usd": 0.0})
        day_bucket["events"] += 1
        day_bucket["cost_usd"] += ev.cost_usd
        if ev.label:
            rpt.by_label[ev.label] = rpt.by_label.get(ev.label, 0) + 1

    served = rpt.total_input_tokens + rpt.total_cache_read_tokens
    rpt.cache_hit_rate = rpt.total_cache_read_tokens / served if served else 0.0
    if rpt.by_day:
        avg_daily = rpt.total_cost_usd / max(len(rpt.by_day), 1)
        rpt.projected_monthly_cost = avg_daily * 30.0
    return rpt


def format_report(rpt: InsightsReport) -> str:
    """Render an :class:`InsightsReport` as a Markdown summary."""
    lines = [
        "# Xerxes Insights",
        "",
        f"Events: {rpt.total_events}",
        f"Total cost: ${rpt.total_cost_usd:.4f}",
        f"Input tokens: {rpt.total_input_tokens:,}",
        f"Output tokens: {rpt.total_output_tokens:,}",
        f"Cache read tokens: {rpt.total_cache_read_tokens:,}",
        f"Cache hit rate: {rpt.cache_hit_rate * 100:.1f}%",
        f"Projected monthly cost: ${rpt.projected_monthly_cost:.2f}",
    ]
    if rpt.by_model:
        lines.append("\n## Top models")
        for model, stats in sorted(rpt.by_model.items(), key=lambda kv: -kv[1]["cost_usd"]):
            lines.append(
                f"- {model}: ${stats['cost_usd']:.4f} ({stats['events']:.0f} events, "
                f"{int(stats['in_tokens']):,} in / {int(stats['out_tokens']):,} out)"
            )
    return "\n".join(lines)


__all__ = ["InsightsReport", "build_report", "format_report"]

File: src/typescript/src/runtime/insights.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

export interface CostEventLike {
  readonly cacheCreationTokens?: number
  readonly cacheReadTokens?: number
  readonly cache_creation_tokens?: number
  readonly cache_read_tokens?: number
  readonly costUsd?: number
  readonly cost_usd?: number
  readonly in_tokens?: number
  readonly inputTokens?: number
  readonly label?: string
  readonly model: string
  readonly out_tokens?: number
  readonly outputTokens?: number
  readonly timestamp: string
}

export interface ModelInsights {
  readonly costUsd: number
  readonly events: number
  readonly inputTokens: number
  readonly outputTokens: number
}

export interface DayInsights {
  readonly costUsd: number
  readonly events: number
}

export interface InsightsReport {
  readonly byDay: Readonly<Record<string, DayInsights>>
  readonly byLabel: Readonly<Record<string, number>>
  readonly byModel: Readonly<Record<string, ModelInsights>>
  readonly cacheHitRate: number
  readonly projectedMonthlyCost: number
  readonly totalCacheCreationTokens: number
  readonly totalCacheReadTokens: number
  readonly totalCostUsd: number
  readonly totalEvents: number
  readonly totalInputTokens: number
  readonly totalOutputTokens: number
}

export interface BuildInsightsOptions {
  readonly days?: number
  readonly now?: Date | (() => Date)
}

/** Aggregate cost events into model, day, label, cache, and burn-rate insights. */
export function buildInsightsReport(
  events: Iterable<CostEventLike>,
  options: BuildInsightsOptions = {},
): InsightsReport {
  const now = resolveNow(options.now)
  const days = options.days
  if (days !== undefined && (!Number.isFinite(days) || days < 0)) throw new RangeError('days must be non-negative')
  const horizon = days && days > 0 ? new Date(now.valueOf() - days * 86_400_000) : undefined
  const byModel: Record<string, MutableModelInsights> = {}
  const byDay: Record<string, MutableDayInsights> = {}
  const byLabel: Record<string, number> = {}
  let totalEvents = 0
  let totalCostUsd = 0
  let totalInputTokens = 0
  let totalOutputTokens = 0
  let totalCacheReadTokens = 0
  let totalCacheCreationTokens = 0

  for (const event of events) {
    const timestamp = parseTimestamp(event.timestamp, now)
    if (horizon !== undefined && timestamp.valueOf() < horizon.valueOf()) continue
    const costUsd = numberValue(event.costUsd ?? event.cost_usd)
    const inputTokens = numberValue(event.inputTokens ?? event.in_tokens)
    const outputTokens = numberValue(event.outputTokens ?? event.out_tokens)
    const cacheReadTokens = numberValue(event.cacheReadTokens ?? event.cache_read_tokens)
    const cacheCreationTokens = numberValue(event.cacheCreationTokens ?? event.cache_creation_tokens)
    totalEvents += 1
    totalCostUsd += costUsd
    totalInputTokens += inputTokens
    totalOutputTokens += outputTokens
    totalCacheReadTokens += cacheReadTokens
    totalCacheCreationTokens += cacheCreationTokens

    const model = event.model
    const modelBucket = byModel[model] ?? { events: 0, costUsd: 0, inputTokens: 0, outputTokens: 0 }
    modelBucket.events += 1
    modelBucket.costUsd += costUsd
    modelBucket.inputTokens += inputTokens
    modelBucket.outputTokens += outputTokens
    byModel[model] = modelBucket

    const day = timestamp.toISOString().slice(0, 10)
    const dayBucket = byDay[day] ?? { events: 0, costUsd: 0 }
    dayBucket.events += 1
    dayBucket.costUsd += costUsd
    byDay[day] = dayBucket

    if (event.label) byLabel[event.label] = (byLabel[event.label] ?? 0) + 1
  }

  const served = totalInputTokens + totalCacheReadTokens
  const projectedMonthlyCost = Object.keys(byDay).length
    ? totalCostUsd / Object.keys(byDay).length * 30
    : 0
  return Object.freeze({
    totalEvents,
    totalCostUsd,
    totalInputTokens,
    totalOutputTokens,
    totalCacheReadTokens,
    totalCacheCreationTokens,
    byModel: freezeModelBuckets(byModel),
    byDay: freezeDayBuckets(byDay),
    byLabel: Object.freeze({ ...byLabel }),
    cacheHitRate: served ? totalCacheReadTokens / served : 0,
    projectedMonthlyCost,
  })
}

/** Source-compatible concise alias. */
export const buildReport = buildInsightsReport

/** Render an insights report as the source Markdown overview. */
export function formatInsightsReport(report: InsightsReport): string {
  const lines = [
    '# Xerxes Insights',
    '',
    'Events: ' + report.totalEvents,
    'Total cost: $' + report.totalCostUsd.toFixed(4),
    'Input tokens: ' + formatInteger(report.totalInputTokens),
    'Output tokens: ' + formatInteger(report.totalOutputTokens),
    'Cache read tokens: ' + formatInteger(report.totalCacheReadTokens),
    'Cache hit rate: ' + (report.cacheHitRate * 100).toFixed(1) + '%',
    'Projected monthly cost: $' + report.projectedMonthlyCost.toFixed(2),
  ]
  const models = Object.entries(report.byModel)
    .sort(([, left], [, right]) => right.costUsd - left.costUsd)
  if (models.length) {
    lines.push('\n## Top models')
    for (const [model, stats] of models) {
      lines.push('- ' + model + ': $' + stats.costUsd.toFixed(4)
        + ' (' + stats.events + ' events, ' + formatInteger(stats.inputTokens)
        + ' in / ' + formatInteger(stats.outputTokens) + ' out)')
    }
  }
  return lines.join('\n')
}

/** Source-compatible concise alias. */
export const formatReport = formatInsightsReport

interface MutableModelInsights {
  costUsd: number
  events: number
  inputTokens: number
  outputTokens: number
}

interface MutableDayInsights {
  costUsd: number
  events: number
}

function resolveNow(value: Date | (() => Date) | undefined): Date {
  const now = typeof value === 'function' ? value() : value ?? new Date()
  if (!(now instanceof Date) || Number.isNaN(now.valueOf())) throw new RangeError('now must be a valid Date')
  return now
}

function parseTimestamp(value: string, fallback: Date): Date {
  const parsed = new Date(value)
  return Number.isNaN(parsed.valueOf()) ? fallback : parsed
}

function numberValue(value: unknown): number {
  return typeof value === 'number' && Number.isFinite(value) ? value : 0
}

function freezeModelBuckets(values: Readonly<Record<string, MutableModelInsights>>): Readonly<Record<string, ModelInsights>> {
  const result: Record<string, ModelInsights> = {}
  for (const [key, value] of Object.entries(values)) result[key] = Object.freeze({ ...value })
  return Object.freeze(result)
}

function freezeDayBuckets(values: Readonly<Record<string, MutableDayInsights>>): Readonly<Record<string, DayInsights>> {
  const result: Record<string, DayInsights> = {}
  for (const [key, value] of Object.entries(values)) result[key] = Object.freeze({ ...value })
  return Object.freeze(result)
}

function formatInteger(value: number): string {
  return value.toLocaleString('en-US')
}
